In [6]:
import pandas as pd

In [11]:
file_path = r"C:\Users\samba\OneDrive\Desktop\Data_Engineer\retail_store_sales_pipeline_project\01. data/retail_store_sales.csv"

In [12]:
df = pd.read_csv(file_path)

In [13]:
print(df.shape)

(12575, 11)


In [14]:
print(df.columns)

Index(['Transaction ID', 'Customer ID', 'Category', 'Item', 'Price Per Unit',
       'Quantity', 'Total Spent', 'Payment Method', 'Location',
       'Transaction Date', 'Discount Applied'],
      dtype='object')


In [19]:
df.dtypes

Transaction ID       object
Customer ID          object
Category             object
Item                 object
Price Per Unit      float64
Quantity            float64
Total Spent         float64
Payment Method       object
Location             object
Transaction Date     object
Discount Applied     object
dtype: object

In [18]:
df.head(10)

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,2024-04-08,True
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,2023-07-23,True
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,2022-10-05,False
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,NaN
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,2022-10-02,False
5,TXN_7482416,CUST_09,Patisserie,NaN,NaN,10.0,200.0,Credit Card,Online,2023-11-30,NaN
6,TXN_3652209,CUST_07,Food,Item_1_FOOD,5.0,8.0,40.0,Credit Card,In-store,2023-06-10,True
7,TXN_1372952,CUST_21,Furniture,NaN,33.5,NaN,NaN,Digital Wallet,In-store,2024-04-02,True
8,TXN_9728486,CUST_23,Furniture,Item_16_FUR,27.5,1.0,27.5,Credit Card,In-store,2023-04-26,False
9,TXN_2722661,CUST_25,Butchers,Item_22_BUT,36.5,3.0,109.5,Cash,Online,2024-03-14,False


In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Transaction Date  12575 non-null  object 
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


In [21]:
df.describe(include = 'all')

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
count,12575,12575,12575,11362,11966.000000,11971.000000,11971.000000,12575,12575,12575,8376
unique,12575,25,8,200,NaN,NaN,NaN,3,2,1114,2
top,TXN_6867343,CUST_05,Furniture,Item_2_BEV,NaN,NaN,NaN,Cash,Online,2022-05-30,True
freq,1,544,1591,126,NaN,NaN,NaN,4310,6354,26,4219
mean,NaN,NaN,NaN,NaN,23.365912,5.536380,129.652577,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,10.743519,2.857883,94.750697,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,5.000000,1.000000,5.000000,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,14.000000,3.000000,51.000000,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,23.000000,6.000000,108.500000,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,33.500000,8.000000,192.000000,NaN,NaN,NaN,NaN


In [22]:
df.isnull().sum()

Transaction ID         0
Customer ID            0
Category               0
Item                1213
Price Per Unit       609
Quantity             604
Total Spent          604
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64

In [23]:
df.duplicated().sum()

np.int64(0)

----------------------------------------------------------------------------------------
Column role Identification
----------------------------------------------------------------------------------------

These are your current numerical columns:

- Price Per Unit
- Quantity
- Total Spent
- Categorical columns

These are your categorical columns:

- Transaction ID
- Customer ID
- Category
- Item
- Payment Method
- Location
- Discount Applied: after cleaning this should become boolean/category, not plain object
Date/time column
- Transaction Date

It is currently stored as text/object, but it parses cleanly into a date.

- Unique identifier column
- Transaction ID

Reason:

every row has a value
all values are unique
no duplicates found

----------------------------------------------------------------------------------------
What is Dirty in this dataset?
----------------------------------------------------------------------------------------

This dataset is not dirty in the “ERROR / UNKNOWN / random symbols” way.
It is dirty in a more realistic retail-operations way: missing fields, wrong storage types, and incomplete transactions.

What I found

A. Missing item names
Item has 1213 nulls

* Problem:

hard to do product-level analysis
may need imputation if inferable, otherwise keep as "Unknown Item" or leave null depending on layer purpose

B. Missing numeric values
Price Per Unit has 609 nulls
Quantity has 604 nulls
Total Spent has 604 nulls

* Problem:

these directly affect revenue analytics
they also affect ML target and features

C. Missing discount flag
Discount Applied has 4199 nulls

* Problem:

this is probably a yes/no field, but many rows are blank
we need to decide how to treat blanks:
unknown discount status
or infer False

For this dataset, I would not silently assume all blanks are False until we inspect relationships more carefully in Part 2.

D. Wrong data types
Transaction Date is object, should become datetime
Discount Applied is object, should become boolean
Quantity is float, but logically it is an integer sales count

E. Incomplete transactions

Some rows are partially missing but still present, for example:

Item missing but Total Spent present
Quantity and Total Spent missing together
Price Per Unit missing but other parts present

These rows need a clear Silver strategy.

----------------------------------------------------------------------------------------
Extra quality checks i ran
----------------------------------------------------------------------------------------

Date parsing
All Transaction Date values parsed successfully 
date range is:
2022-01-01 to 2025-01-18

So the dataset covers roughly 3 years of transactions.

Total consistency

I checked whether:

Total Spent == Price Per Unit * Quantity

for rows where all 3 values are present.

Result:

0 mismatches

This is excellent.

It means:

- Total Spent is trustworthy
- missing Total Spent can likely be recomputed when both price and quantity exist
- missing Price Per Unit can be derived if total and quantity exist
- missing Quantity can be derived if total and price exist and division is exact
- Category-item consistency

I checked whether one item appears under multiple categories.

Result:

0 items mapped to multiple categories

This is also excellent.

That means:

- Item can likely be imputed from Category only in a weak way, but not uniquely
- Category is stable and trustworthy
- Duplicate transaction IDs
- 0 duplicate transaction IDs

----------------------------------------------------------------------------------------
Key business columns for the project
----------------------------------------------------------------------------------------

Key numerical columns

These will drive analytics and ML:

Price Per Unit
Quantity
Total Spent
Key categorical columns

These will drive slicing/grouping in SQL and Power BI:

- Category
- Item
- Payment Method
- Location
- Discount Applied
- Customer ID
- Date column
- Transaction Date
- Identifier column
- Transaction ID

----------------------------------------------------------------------------------------
What should Bronze look like?
----------------------------------------------------------------------------------------

Because Bronze must keep the raw file as-is, your Bronze table should use the exact raw columns:

- Transaction ID
- Customer ID
- Category
- Item
- Price Per Unit
- Quantity
- Total Spent
- Payment Method
- Location
- Transaction Date
- Discount Applied

Bronze rule for this project

Store the CSV exactly as uploaded:
- same column names
- same missing values
- no cleaning
- no renaming yet

----------------------------------------------------------------------------------------
What cleaning will be needed for Silver?
----------------------------------------------------------------------------------------

Based on the inspection, Silver should do these things:

Required Silver cleaning steps

1. Standardize column types
- convert Transaction Date → datetime
- convert Discount Applied → boolean
- convert Quantity → integer where valid

2. Handle missing numeric values

Use business-safe rules:

- if Price Per Unit and Quantity exist but Total Spent is null
→ compute Total Spent = Price Per Unit * Quantity
- if Total Spent and Quantity exist but Price Per Unit is null
→ compute Price Per Unit = Total Spent / Quantity
- if Total Spent and Price Per Unit exist but Quantity is null
→ compute Quantity = Total Spent / Price Per Unit
- only if it gives a valid whole-number quantity

3. Handle missing item values

Possible strategy:

- keep null as "Unknown Item" for analytics friendliness
- preserve original missingness in a flag column if needed

4. Handle missing discount values

Best Silver approach:

- convert True/False properly
- keep missing as NULL first
- optionally add a derived label later such as Unknown

5. Validate totals

Recheck:

- Total Spent = Price Per Unit × Quantity

after cleaning

6. Add data-quality flags

This dataset is a good candidate for flags like:

- item_missing_flag
- discount_missing_flag
- imputed_price_flag
- imputed_quantity_flag
- imputed_total_flag


----------------------------------------------------------------------------------------
What Gold analytics tables make sense?
----------------------------------------------------------------------------------------

This dataset is retail-focused, so your Gold layer should feel more business/owner oriented than the cafe project.

Strong Gold tables for Power BI

1. gold.daily_sales_summary

Use for:

- daily revenue
- order count
- average order value
- quantity sold

2. gold.product_performance

Use for:

- best-selling items
- highest-revenue items
- average unit price by item

3. gold.category_analysis

Use for:

- revenue by category
- quantity by category
- order contribution by category

This is more important here than in your cafe project.

4. gold.payment_analysis

Use for:

- payment method distribution
- revenue by payment type
- order count by payment type

5. gold.location_analysis

Use for:

- online vs in-store comparison
- revenue split by channel
- average basket value by location

6. gold.monthly_trends

Use for:

- month-over-month revenue
- seasonal trend
- order volume trend

7. gold.discount_analysis

Use for:

- performance of discounted vs non-discounted transactions
- average order value with discount vs without
- discount usage rate

This is very retail-specific and will make this project stand out.

----------------------------------------------------------------------------------------
What ML target makes the most sense?
----------------------------------------------------------------------------------------

For this dataset, the best target is likely:

Recommended ML target
- Daily total revenue

That means we aggregate transaction-level data into daily features, then predict:
- daily_revenue

Why this is the best target
- it is business-relevant
- it aligns naturally with time-series style forecasting
- it is easy to connect to Power BI
- it matches the kind of “store owner” questions people care about:
- how much will we sell tomorrow?
- next week?
- next month?
- Possible ML feature groups

From this dataset, future gold.ml_features can include:

- Date features
- year
- month
- day
- day_of_week
- weekend_flag
- Sales behavior features
- daily_order_count
- daily_quantity_sold
- avg_order_value
- avg_price_per_unit
- Mix features
- online_order_ratio
- discount_usage_ratio
- category diversity per day
- item diversity per day
- Time-lag features
- lag_1
- lag_7
- rolling_avg_7

This can be more retail-specific than your first project.

----------------------------------------------------------------------------------------
My recommendation for column roles going forward
----------------------------------------------------------------------------------------

Identifier
- Transaction ID

Date
- Transaction Date

Core business measures
- Price Per Unit
- Quantity
- Total Spent

Main retail dimensions
- Category
- Item
- Payment Method
- Location
- Discount Applied

Secondary dimension
- Customer ID